Added sorting by modified time in Folder Util.

Tzaoh opened the pull request:
Initially Folder class does not allow to sort by Modified Time.
I added some logic to accomplish it. The PR is backwards with previous versions calls.

If someone is trying to sort with `$sort = true` it will sort by name.
If someone is trying to sort with `$sort = 'name'` it will sort by name.
If someone is trying to sort with `$sort = 'time'` it will sort by modified time.
If someone is trying to sort with `$sort = false` it will not sort.

Added method test in FolderTest.php

Commit:
added ordering by modified time in Folder Util.

Now the results returned by Folder::path can be ordered by file/dir
modification time.

Tzaoh commented:
Aparently in travis is failing two methods but not in my local windows. Any ideas?

chinpei215 commented:
Does `sleep(1)` solve it?

chinpei215 commented on src/Filesystem/Folder.php:
```diff
@@ -194,16 +194,32 @@ public function read($sort = true, $exceptions = false, $fullPath = false)
             }
         }
+
         if ($sort || $this->sort) {
```
Doesn't `$this->sort` support new option?

chinpei215 commented on src/Filesystem/Folder.php:
```diff
@@ -194,16 +194,32 @@ public function read($sort = true, $exceptions = false, $fullPath = false)
+                $dirs[$item->getCTime()][] = $name;
             } else {
-                $files[] = $name;
+                $files[$item->getCTime()][] = $name;
```
Ah probably you changed the current behavior. `$item->getCTime()` should be called only if `$sort` is `true`.

markstory replied:
Won't this have issues if there are two files with the same modified time?

Tzaoh replied:
Nope, [$item->getCTime()][] they are added as a subarray

chinpei215 replied:
I don't think applying `sort` function against 2d array works as you expected. For example Can your commit keep the same order if their are 3 files like the following:
```
ctime   name
100     a.txt
100     c.txt
101     b.txt
```

Tzaoh replied:
@chinpei215 hey :)
$item->getCTime() is always called but the `korder` call is called depending on the sort chosen.

Apart from that, imo I would consider correct:
```
ctime name
100   a.txt
100   c.txt
101   b.txt
```
and
```
ctime name
100   c.txt
100   a.txt
101   b.txt
```

Because `a.txt` and `c.txt` has the same `ctime` and we are not sorting by time and name, only but time. I created another test that accomplish it and change some gaps.

chinpei215 replied:
@Tzaoh I am not sure that you get what I mean. Sorry if you have already got it. I wanted to say, the `sort($files)` seems not to work correctly. Isn't `$files` a 2d array?
```php
$files = [
    100 => [ 'a.txt', 'c.txt' ],
    101 => [ 'b.txt', ],
]
```

Tzaoh replied:
No problem mate. Thanks for clearing it to me.
Actually I am having the correct structure (I attach a photo). I guess you have the last patch, right?

![image](https://cloud.githubusercontent.com/assets/5722357/14935860/fbb8e7f0-0ee0-11e6-89dd-1d035fc8dd60.png)

The two errors I am getting in travis shows that the result have different order, quite surprising because in my local test all are correct. Maybe @markstory could point me in the right way about why this.

Thanks both of you

chinpei215 replied:
@Tzaoh I think `array_merge()` at line 230 makes the *crrect* structure. But `sort()` is called at line 221.

Tzaoh replied:
I understood what you wanted to tell me, thought you were talking about ::Time sorting not ::Name.
Lets see about these tests. They are still passed on my local win.

markstory commented on src/Filesystem/Folder.php:
```diff
@@ -194,16 +194,32 @@ public function read($sort = true, $exceptions = false, $fullPath = false)
         if ($sort || $this->sort) {
-            sort($dirs);
-            sort($files);
+            if ($sort === 'time') {
```
It would be preferable to have constants for the new modes. Having constants helps prevent typos.

Commit:
Consts For Sort Ordering

Commit:
Using Const in method test

Commit:
Correcting some gaps

·  Added a second test for chinpei215 request.
· use "||" instead of "or" in conditionals
· "elseif" instead of "else if"

markstory commented:
Several [tests are failing](https://travis-ci.org/cakephp/cakephp/jobs/126868334) around the modified feature.

Commit:
Changed the way sort are apply

As @chinpei215 pointed there were some errors when SORT_NAME was
provided.

Tzaoh commented:
yep @markstory as @chinpei215 pointed I didnt think in some cases. Lets see if this is okay.

chinpei215 commented:
Looks good to me. Sorry for my bad explanation :cry:
`DirectoryIterator` seems to sort files by the name on windows. That is why the tests passing on your local environment.

Tzaoh commented:
@chinpei215 good to know, thanks for your help :)

dereuromark commented on src/Filesystem/Folder.php:
```diff
@@ -165,7 +183,7 @@ public function cd($path)
      * @return array Contents of current directory as an array, an empty array on failure
      */
-    public function read($sort = true, $exceptions = false, $fullPath = false)
+    public function read($sort = self::SORT_NAME, $exceptions = false, $fullPath = false)
```
You also need to update the doc block line to `string|bool`.

dereuromark commented on src/Filesystem/Folder.php:
```diff
@@ -183,6 +201,12 @@ public function read($sort = true, $exceptions = false, $fullPath = false)
             return [$dirs, $files];
         }
 
+        if (!is_bool($sort) && array_key_exists($sort, $this->_fsorts)) {
```
IMO isset() should be preferred over array_key_exists() as null values are not relevant here.

dereuromark commented on src/Filesystem/Folder.php:
```diff
@@ -194,16 +218,27 @@ public function read($sort = true, $exceptions = false, $fullPath = false)
+            ksort($files);
         }
+
+        if (!empty($dirs)) {
```
Should be either `!$dirs` or `count(...) === 0`.
!empty() cloaks spelling mistakes.

dereuromark commented on src/Filesystem/Folder.php:
```diff
@@ -76,6 +86,14 @@ class Folder
     public $mode = 0755;
 
     /**
+     *
```
type array missing

Commit:
Docs and improvements added. Typos corrected.

Tzaoh commented:
· Shouldnt be better to use `rename` to move files instead of copy && delete?
· What about adding in the future to return more data, not just the name/fullname but other data like mod time, created time...
This last proposition will break some test and will not be backward compatible as the result data would be an array (or another type of data)

markstory commented:
> Shouldnt be better to use php rename() to move files instead of copy && unlink?

Yes, but please do a separate pull request for that.

markstory commented:
>  This last proposition will break some test and will not be backward compatible as the result data would be an array (or another type of data)

Returning `File` objects would be preferable to arrays.

ravage84 commented:
Can this be backported to 2.x :question:

dereuromark commented:
Sure, someone just needs to make the PR for 2.8 branch.
